### PR TITLE
#162949042 Users should be able to comment on articles

### PR DESCRIPTION
**Pull request by johnwayodi**
#### What does this PR do?
Implement frontend functionality for commenting on articles
#### Description of Task to be completed?
- [x] setup actions, reducers, sagas
- [x] implement add comment feature
- [x] implement get all comments feature
- [x] implement update comment feature
- [x] implement delete comment feature
- [x] write tests

1. Navigate to a folder where you would prefer to test the repository.

2. Create a directory
`mkdir myproject`

3. Navigate to the directory "myproject"
`cd myproject`

4. Clone the author's haven repository
`git clone https://github.com/andela/ah-robotics-frontend.git`

5. Change the directory into the "ah-robotics-fronted"
`cd ah-robotics-frontend`

6.  Checkout to the branch ft-comments-crud-162949042
`git checkout ft-comments-crud-162949042`

7. Install dependencies from the package.json
 `npm install` to install all the modules listed in the `package.json` file.

8. Start the application in development mode
`npm start`
> The application opens http://localhost:3000 in the browser successfully if there are no issues.

 #### Test out the feature
The application can be accessed on heroku using the URL below : 
[https://ah-robotics-frontend-sta-pr-15.herokuapp.com](https://ah-robotics-frontend-sta-pr-15.herokuapp.com);

9. Register a new user by heading over to the **register** page.
10. After Registering a user, head over to the **login** page and use the details that you registered with to login

11. Select any article from the articles list in order to load the article view and access the comment section of the article.

12. Type a comment in the text area and then press the **Add Comment** button 
13. When you add a comment you can also **_Update_** it or **_Delete_** it as the options will appear below the comment text.

###  Running Tests
14. Add a **.env** file to the root folder of the project and add the configurations for social authentication in the following format, this is necessary for social login tests to pass.

           REACT_APP_API_KEY=sample_key
           REACT_APP_AUTH_DOMAIN=sample_domain
> **NOTE** : The values of the above variables are in the pinned messages section on the slack group.
15. Run the following command to make sure all tests are passing:
`npm test`

#### Any background context you want to provide?
None
#### What are the relevant pivotal tracker stories?

[#162949042](https://www.pivotaltracker.com/story/show/162949042)

### Screenshots
1. Adding a comment
image

2. Updating a comment
image

3. Deleting a comment
image

### CODE REVIEW

File: src/components/Comments/comment.scss [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  background: rgba(0, 0, 0, 0);
+  border: none;
+  outline: 0;
+  cursor: text;
+  resize: none;
+  border-bottom: solid 1px rgba(230, 230, 230, 0.658);
+}
+
+#comment-section {
+  //background-color: red;
```
Comment (jomasim) [inline src/components/Comments/comment.scss]: remove this commented line

File: src/components/Comments/comment.scss [hunk 2]
```diff
@@ -0,0 +1,44 @@
+$primary-color: #00d0a0;
+#comment-button {
+  background-color: $primary-color !important;
+  color: white !important;
+  border-radius: 20px !important;
+  //padding: 5px !important;
```
Comment (jomasim) [inline src/components/Comments/comment.scss]: remove commented line

Comment (Teatoller) [reply src/components/Comments/comment.scss]: Hello, Wayodi.
I am able to get my comments posted.
A small glitch though is that I have to reload the page after clicking on **Add Comment** button for the circle to be complete.

Regards
Steven Ennis

**Comment by sharkdevs** [comment]
Hello @johnwayodi, there is a problem once a user posts a comment. There should be a loader to show the progress of the posting. After a comment is posted successfully, kindly consider clearing the input field. Lastly, ensure you have a way to tell the user that their comment was successfully posted. I did not know that my comment was posted. I ended up clicking the comment button severally and this was the result after i refreshed the page:
screen shot 2019-02-25 at 09 50 36

screen shot 2019-02-25 at 09 46 08

**Comment by Paulvitalis200** [comment]
Hey @johnwayodi ,

I have managed to test this feature and it works. However let us add a loader for it to be more effective

File: src/components/Comments/comment.scss (hunk 2 above)
Comment (johnwayodi) [reply src/components/Comments/comment.scss]: > A small glitch though is that I have to reload the page after clicking on **Add Comment** button for the circle to be complete.

@Teatoller , I have made a few adjustments to tackle this. you can try again and see if it's okay.

Comment (johnwayodi) [reply src/components/Comments/comment.scss]: @jomasim I have made the changes. thanks for reviewing

**Comment by johnwayodi** [comment]
> Hello @johnwayodi, there is a problem once a user posts a comment. There should be a loader to show the progress of the posting. After a comment is posted successfully, kindly consider clearing the input field. Lastly, ensure you have a way to tell the user that their comment was successfully posted. I did not know that my comment was posted. I ended up clicking the comment button severally...

@sharkdevs and @Paulstar200 , I have made the changes and added the loader, you can review again and see if its okay

**Comment by jomasim** [comment]
When you have one comment it should display:
``` 1 comment ```
screenshot 2019-02-27 at 14 27 48

**Comment by johnwayodi** [comment]
> When you have one comment it should display:
> `1 comment`

@jomasim  working on it. thanks for reviewing

**Review by waracci — COMMENTED** [review]
Hi @johnwayodi , I have been able to go through your work and I noticed that if I comment on a specific article, the comments for the previously commented on article are showing on all the other articles. Kindly look into this. You should consider clearing the comments state while opening another article.

**Comment by johnwayodi** [comment]
> Hi @johnwayodi , I have been able to go through your work and I noticed that if I comment on a specific article, the comments for the previously commented on article are showing on all the other articles. Kindly look into this. You should consider clearing the comments state while opening another article.

Thanks, @waracci  for the review, am working on it.

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @johnwayodi ,
This is awesome work. Everything works as expected. The only thing you can improve on is if there is only one comment, it should say 1 comment instead of 1 comments.

**Comment by johnwayodi** [comment]
> Hey @johnwayodi ,
> This is awesome work. Everything works as expected. The only thing you can improve on is if there is only one comment, it should say 1 comment instead of 1 comments.

thanks for reviewing @Paulstar200, @jomasim had also noted the same issue, I have worked on it. Checkout the new implementation.

**Comment by johnwayodi** [comment]
> Hi @johnwayodi , I have been able to go through your work and I noticed that if I comment on a specific article, the comments for the previously commented on article are showing on all the other articles. Kindly look into this. You should consider clearing the comments state while opening another article.

@waracci I have worked on this. Thanks for reviewing.

File: src/redux/actions/CommentActions/actions.test.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    expect(actions.updateCommentFail({}).type).toEqual(types.UPDATE_COMMENT_FAIL);
+  });
+});
+
+
+describe('Delete Comment Actions Tests', () => {
+  it('should dispatch DELETE_COMMENT_START', () => {
+    expect(actions.deleteArticleComment({}).type).toEqual(types.DELETE_COMMENT_START);
+  });
+  it('should dispatch LOGIN_ERROR', () => {
```
Comment (jomasim) [inline src/redux/actions/CommentActions/actions.test.js]: rename this test to DELETE comment type

File: src/redux/actions/CommentActions/actions.test.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    expect(actions.deleteCommentSuccess({}).type).toEqual(types.DELETE_COMMENT_SUCCESS);
+  });
+  it('should dispatch LOGIN_SUCCESS', () => {
```
Comment (jomasim) [inline src/redux/actions/CommentActions/actions.test.js]: rename this test to DELETE comment type

File: src/redux/actions/CommentActions/actions.test.js (hunk 3 above)
Comment (johnwayodi) [reply src/redux/actions/CommentActions/actions.test.js]: I have renamed this correctly

File: src/redux/actions/CommentActions/actions.test.js (hunk 4 above)
Comment (johnwayodi) [reply src/redux/actions/CommentActions/actions.test.js]: I have renamed this correctly, thanks @jomasim for reviewing

**Pull request merged**